polars/polars-time/src/chunkedarray/utf8/infer.rs

```diff
@@ -10,6 +12,104 @@ use crate::chunkedarray::utf8::patterns::Pattern;
 use crate::chunkedarray::utf8::strptime;
 use crate::prelude::utf8::strptime::StrpTimeState;
 
+const ALLOWED_SUBSTRINGS: [&str; 41] = [
+    "jan",
+    "feb",
+    "mar",
+    "apr",
+    "may",
+    "jun",
+    "jul",
+    "aug",
+    "sep",
+    "oct",
+    "nov",
+    "dec",
+    "t",
+    "january",
+    "february",
+    "march",
+    "april",
+    "june",
+    "july",
+    "august",
+    "september",
+    "october",
+    "november",
+    "december",
+    "am",
+    "pm",
+    "z",
+    "mon",
+    "tue",
+    "wed",
+    "thu",
+    "fri",
+    "sat",
+    "sun",
+    "monday",
+    "tuesday",
+    "wednesday",
+    "thursday",
+    "friday",
+    "saturday",
+    "sunday",
+];
+
+fn is_not_inferable_datetime(val: &str) -> bool {
+    // Return true if any of the following:
+    // - val is over 50 characters long
+    // - val contains a substring other than those in `ALLOWED_SUBSTRINGS`
+    // - val contains a sequence of more than 9 consecutive digits
+    // - val contains any character which is not:
+    //   - a digit
+    //   - a letter
+    //   - one of ' ', '/', '-', '.', '+', ':'
+    // - val contains fewer than 6 digits in total
+    // - val contains more than 30 digits in total
+    //   (note that '200000-01-01T01:00:00.123456789+01:00' is valid)
+    // - val doesn't have at least 4 consecutive digits
+    //
+    // Note: this is not complete, but it's just intended to
+    // provide a fail-fast mechanism if a string is not going to
+    // be automatically parseable anyway.
+    let mut total_digits: u8 = 0;
+    let mut consecutive_digits: u8 = 0;
+    let mut max_consecutive_digits: u8 = 0;
+    let mut current_substring = String::new();
+    if val.len() > 50 {
+        return true;
+    }
+    for character in val.chars() {
```

Comment:
There are a few things we need to be careful here. 

1. I think this function should only be called if the first pattern has failed. So that the happy path (where the first pattern always matches goes on without interference).

2. We should be careful with the string allocations here. I believe we should be able to do this with indexing and adding the allowed capital case variants as well. We could put smaller search arrays in a lookup table where we lookup the subslice by first character.

3. I think we should have such a function per pattern type group. This can make each function more opinionated and thus faster in many cases.